keps/sig-auth/3926-handling-undecryptable-resources/README.md

```diff
@@ -0,0 +1,955 @@
+by one or more prompts from the `kubectl` client. In case of a plain HTTP DELETE
+request, at least issue a warning about what the impact of the action might be.
+Gate with an additional admission layer.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+### Current state
+
+The encryption/decryption for encryption at rest is implemented via transformers
+that get applied to a resource in code that handles resource read/write from etcd3
+databases.
+
+The storage handling does not change with KMSv2, a resource transformer is provided
+in that case, too.
+
+References:
+1. [Encryption transformers creation](https://github.com/kubernetes/kubernetes/blob/8decaf3ae7f410ab3f3774f3895b9f3124b8a4c6/staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go#L185)
+2. [Example of a resource retrieval with the tranformer](https://github.com/kubernetes/kubernetes/blob/8decaf3ae7f410ab3f3774f3895b9f3124b8a4c6/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L155-L157)
+
+The code example 2. above shows that currently, when reading a resource fails, we
+lose all the context about the resource and a non-wrapping, generic internal error
+is returned.
+
+#### Retrieving a failing resource
+
+The current API errors don't appear to include an error status specific to storage. Therefore
+a new status should be introduced - `StatusReasonStoreReadError`.
+
+```go
+  // StatusReasonStoreReadError means that the server encountered an error while
+  // retrieving resources from the backend object store.
+  // This may be due to backend database error, or because processing of the read
+  // resource failed.
+  // Details:
+  //   "kind" string - the kind attribute of the resource being acted on.
+  //   "name" string - the prefix where the reading error(s) occurred
+  //   "causes" []StatusCause 
+  //      - (optional):
+  //        - "type" CauseType - CauseTypeUnexpectedServerResponse
+  //        - "message" string - the error message from the store backend
+  //        - "field" string - the full path with the key of the resource that failed reading
+  //
+  // Status code 500
+  StatusReasonStoreReadError StatusReason = "StorageReadError"
+```
+
+This error will also include full paths to the resources that cannot be read in
+the unstructured, human-readable message.
+
+With this error, a new `ListOption` can be introduced - `IgnoreStorageReadError`.
+
+```go
+type ListOptions struct {
+  ...
+  // IgnoreStorageReadError ignores errors that might occur during reading a
+  // resource from the storage backend. Instead, the message of that error gets
+  // transformed into a warning.
+  // +optional
+  IgnoreStorageReadError bool
```

Comment:
it also makes the resourceVersion in the returned list invalid, and would break in combination with watch

if there are storage transformation errors, I think the critical problem to solve is to surface the specific resources that cannot be read

Reply:
I agree - having such option is asking for very subtle bugs.
In my opinion we need two things:
- expose the information what objects cannot be read
- allow for deleting those objects

Reply:
I was expecting a use-case as described in https://github.com/kubernetes/enhancements/pull/3927#discussion_r1151781712, typically a controller that injects content into a resource, cluster-wide, based on its annotation/label.

I haven't thought of the resourceVersion on lists, though.

It seems that such an option might bring more issues than value. While it was an interesting thought, I agree that it's probably  undesirable at this moment.